File: backend/src/Service/todo_service.py
# ===== AI GENERATED: TodoService =====
# Purpose: Business orchestration for Todo use-cases
# Inputs: repository implementing IToDoRepository
# Returns: DTOs consumed by controllers
# Flow:
# 1. Delegate to repository, handle simple validations

from typing import List, Optional
from ..Contract.i_todo_repository import IToDoRepository
from ..DTO.todo_dto import TodoCreate, TodoUpdate, TodoOut


class TodoService:
    def __init__(self, repository: IToDoRepository):
        self.repository = repository

    def list_todos(self) -> List[TodoOut]:
        return self.repository.list_all()

    def get_todo(self, todo_id: int) -> Optional[TodoOut]:
        return self.repository.get(todo_id)

    def create_todo(self, payload: TodoCreate) -> TodoOut:
        if not payload.title or not payload.title.strip():
            raise ValueError("title is required")
        return self.repository.create(payload)

    def update_todo(self, todo_id: int, payload: TodoUpdate) -> Optional[TodoOut]:
        return self.repository.update(todo_id, payload)

    def delete_todo(self, todo_id: int) -> bool:
        return self.repository.delete(todo_id)

File: frontend/src/services/todoService.js
// ===== AI GENERATED: todoService =====
// Purpose: Axios wrapper for Todo CRUD
// Inputs: uses VITE_API_BASE_URL
// Returns: functions: listTodos, getTodo, createTodo, updateTodo, deleteTodo
// Flow:
// 1. Create axios instance
// 2. Export CRUD helpers

import axios from 'axios'

const baseURL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000'
const api = axios.create({ baseURL })

export async function listTodos() {
  const res = await api.get('/todos/')
  return res.data
}

export async function getTodo(id) {
  const res = await api.get(`/todos/${id}`)
  return res.data
}

export async function createTodo(payload) {
  const res = await api.post('/todos/', payload)
  return res.data
}

export async function updateTodo(id, payload) {
  const res = await api.put(`/todos/${id}`, payload)
  return res.data
}

export async function deleteTodo(id) {
  const res = await api.delete(`/todos/${id}`)
  return res.data
}

export async function checkBackend() {
  const res = await api.get('/health')
  return res.data
}

export async function checkDb() {
  const res = await api.get('/health/db')
  return res.data
}
